### ISSUE TITLE
Parallel transducers?

### ISSUE THREAD

**Issue by sgaure** [JuliaFolds/Transducers.jl#301]
Just a question, or a suggestion, or a speculation.

I know that there is a parallel reduce and a tcollect, and some such global parallelism, but would it be possible to have a pair of transducers which forks out in multiple threads for the downstream, and then somewhere downstream collects it back into a single thread?  Like

`xf1 |> xf2 |> TFork(24) |> xf3 |> xf4 |> TWeave |> xf5 |> xf6`

I.e. `xf1 |> xf2` and `xf5 |> xf6` run as usual, but `xf3 |> xf4` will spread out to 24 threads running in parallel. They are then interleaved back into a single thread by TWeave, possibly with a different order. 

Clearly, for this to work in a nested fashion (e.g. if xf3 also contains a Fork/Weave pair), the TWeave would have to know the right TFork, so it doesn't weave together separate outside threads.

**Comment by tkf** [comment]
Hmm.... Let me try understanding what you are saying. So, in this scenario,

`xf1 |> xf2 |> TFork(24) |> xf3 |> xf4 |> TWeave() |> xf5 |> xf6`

and

`xf1 |> xf2 |> xf3 |> xf4 |> xf5 |> xf6`

are semantically equivalent but the first one can be faster if you have many CPUs?  If that's the case, what is the motivation to _not_ parallelize `xf1 |> xf2` and `xf5 |> xf6`?

I think the closest thing that exist in Transducers.jl ATM is `TCat`. You can do

`xf1 |> xf2 |> Partition(24) |> TCat(24) |> xf3 |> xf4 |> xf5 |> xf6`

although this runs `xf5 |> xf6` in parallel. It also has a bit high overhead since it joins the tasks for `xf3 |> xf4 |> xf5 |> xf6` every 24 steps.

You can also do

`xf1 |> xf2 |> Partition(24) |> Map(xs -> tcollect(xf3 |> xf4, xs)) |> Cat() |> xf5 |> xf6`

although this allocates an array for each 24 steps. I think it's possible to avoid allocation using `append!!` instead of `tcollect` and `Scan` instead of `Map`.

**Comment by sgaure** [comment]
> semantically equivalent but the first one can be faster if you have many CPUs? If that's the case, what is the motivation to not parallelize xf1 |> xf2 and xf5 |> xf6?

It could be that some of the transducers were not thread-safe, or they require too much temporary memory, so you can't afford it.

**Comment by tkf** [comment]
I see. In that case, the two examples I've commented will be the basic approach.

If the order does not matter, you can also use channels:

```julia
ch1 = Channel(xf1 |> xf2, itr)
ch2 = channel_unordered(xf3 |> xf4, ch1; ntasks = 24)  # multi-threaded
collect(xf5 |> xf6, ch2)
```

**Comment by sgaure** [comment]
That's fine. My idea was to mirror the simple Threads.@threads specification (or openmp parallel for) available for iterators (with static or dynamic load-balancing), with an optional builtin "dethreading" somewhere downstream as a bonus.

**Comment by tkf** [comment]
Which part is fine? Do you mean items being unordered?

> My idea was to mirror the simple `Threads.@threads` specification (or openmp parallel for) available for iterators

I don't understand this part. `Threads.@threads` can only be used with arrays.  Supporting iterators requires a package like [ThreadsX.jl](https://github.com/tkf/ThreadsX.jl) that is based on Transducers.jl and [SplittablesBase.jl](https://github.com/JuliaFolds/SplittablesBase.jl).

**Comment by sgaure** [comment]
It’s fine that it’s possible to parallelize code with various tools. I wondered if it could be done directly with special transducers, but it seems to be easier to do it outside transducers.

**Comment by tkf** [comment]
Yeah, it might be reasonable to automate and improve the `Partition`-based method by something like

> ```julia
> ThreadedXF(xf::Transducer; basesize, ntasks, nbuffers) -> xf′::Transducer
> ```
>
> Parallelize transducer `xf` using intermediate buffers, preserving the order of items.
>
> Each consecutive partition of `basesize` items are processed by transducer `xf` in a separate task by accumulating the input items in a buffer.
>
> **Keyword arguments**
>
> - `basesize::Integer = 1`: The size of consecutive items processed by each task.  This determines the maximum length of each buffer.
> - `ntasks::Integer = nthreads()`: The maximum number of tasks that are spawned by `ThreadedXF`.
> - `nbuffers::Integer = ntasks + 1`: The maximum number of buffers.  Up to `basesize * nbuffers` items are buffered.  Use a larger value to improve thread utilization if the processing time of `xf` varies across inputs.

**Comment by sgaure** [comment]
There are easy ways in julia to parallelize loops, and Threads.@spawn macro inheriting from Cilk is great for various parallellizations, including partitioning of iterators for cutting them up in suitable chunks. 

However, besides loops there is another common parallel construction which can be done by channels, but it has to be coded explicitly. A pipeline, the construction which was adressed by IBMs Cell-processor. I.e. you send your data piece by piece into a series of transformations. Once you have filled the pipeline, the parts run in parallel, even though each data item passes through in a serial fashion. 

Combined with wide pipelines, i.e. parallel pipes, it would be a quite nice functionality. However, transducers are probably not the right thing for this, it would mean much higher overhead than now.

**Comment by tkf** [comment]
Transducers.jl and ThreadsX.jl only impose approximated associativity in the reducing function by default.  This is crucial for a deterministic result from parallel computation.  But I'm not against assuming exact associativity and commutativity so that the execution can happen in out-of-order.  Also, sometimes you don't need determinism.  That's why I've already added `channel_unordered`.  It seems your concern can be addressed by supporting out-of-order execution in `ThreadedXF`(say `ordered::Bool` flag).  My guess is that it'd be much better than using `Channel`s in terms of throughput.

**Comment by tkf** [comment]
#431 implements `NondeterministicThreading` transducer that can be used for parallelizing inner transducers and reducing function but still sequentially process outer transducers and `iterate` of the input iterator.  It's not as generic as `ThreadedXF` that I suggested above but it's still useful in some situations.
